### PR TITLE
Resolve issues arising from geopmpy/geopmdpy static analysis

### LINKED ISSUE #2168: Python dependency requirements are not consistent and may be out of date
The requirements.txt and setup.py dependencies in geopmdpy do not match. This isn't necessarily a problem (e.g., if requirements is meant for a different build/test/run use case from setup.py), but the setup version does look insufficient.

The required dasbus versions do not match. I am aware of a 1.5.0+ requirement since they added a feature to help us with #1606. I'm not sure if our later changes add a 1.6 requirement.

Sphinx is missing from the setup.py requirements, which can result in documentation build failures.

Current geopmdpy requirements.txt:
```
cffi>=1.13.2
dasbus>=1.6
PyGObject>=3.34.0
psutil>=5.4.8
docstring_parser>=0.8.1
sphinx>=4.1.0
sphinx_rtd_theme>=0.5.2
jsonschema>=2.6.0
```

setup.py:
```
                    'cffi>=1.6.0',
                    'setuptools>=39.2.0',
                    'future>=0.17.1',
                    'psutil>=5.6.2',
                    'dasbus>=1.5.0',
                    'jsonschema>=2.6.0'
```

geopmpy requirements.txt:
```
matplotlib>=2.2.2
cycler>=0.10.0
numpy>=1.14.3
setuptools>=39.2.0
natsort>=5.3.2
psutil>=5.4.8
pandas>=0.23.0
tables>=3.4.3
cffi>=1.6.0
pyyaml>=5.1.0
future>=0.17.1
```

setup.py
```
                    'pandas>=0.23.0',
                    'natsort>=5.3.2',
                    'matplotlib>=2.2.2',
                    'cycler>=0.10.0',
                    'tables>=3.4.3',
                    'psutil>=5.4.8',
                    'cffi>=1.6.0',
                    'numpy>=1.14.3',
                    'setuptools>=39.2.0',
                    'pyyaml>=5.1.0',
                    'future>=0.17.1',
                    'geopmdpy'
```

Pandas upgraded to 1.0 a while ago. We were previously blocked from upgrading that dependency since we were concurrently supporting python2 and python3 (and 1.0 is when they droppped python2 support). We are able to upgrade now that we dropped python 2 support. We should probably modify our dependency to be a stable release series since they exist now.

### PR DESCRIPTION
**Pull request by bgeltz**
- Relates to #2256.
- Relates to #2257.
- Fixes #2168.

### CODE REVIEW

File: service/src/PlatformTopo.cpp [hunk 1]
```diff
@@ -56,7 +62,8 @@ int geopm_read_cpuid(void)
 
 namespace geopm
 {
-    const std::string PlatformTopoImp::M_CACHE_FILE_NAME = "/tmp/geopm-topo-cache";
+    const std::string PlatformTopoImp::M_CACHE_FILE_NAME = "/tmp/geopm-topo-cache-" + std::to_string(getuid());
+    const std::string PlatformTopoImp::M_SERVICE_CACHE_FILE_NAME = "/run/geopm-service/geopm-topo-cache";
```
Comment (bgeltz) [inline service/src/PlatformTopo.cpp]: Move to initializer list.

File: service/src/PlatformTopo.cpp [hunk 2]
```diff
@@ -423,31 +430,45 @@ namespace geopm
...
-            }
+        int err = stat(cache_file_name.c_str(), &cache_stat);
+        if (err) {
+            throw Exception("PlatformTopoImp::create_cache(): stat failure:",
+                     errno ? errno : GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+        }
+
+        std::string cmd = "out=" + cache_file_name + ";"
+                          "lscpu -x > $out && "
+                          "chmod 644 $out";
```
Comment (bgeltz) [inline service/src/PlatformTopo.cpp]: Use umask beforehand and avoid chmod call.

File: service/src/PlatformTopo.cpp [hunk 3]
```diff
@@ -423,31 +430,45 @@ namespace geopm
...
-                throw Exception("PlatformTopo::create_cache(): Could not pclose lscpu command: ",
-                         errno ? errno : GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
-            }
+        int err = stat(cache_file_name.c_str(), &cache_stat);
+        if (err) {
+            throw Exception("PlatformTopoImp::create_cache(): stat failure:",
+                     errno ? errno : GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+        }
+
+        std::string cmd = "out=" + cache_file_name + ";"
```
Comment (bgeltz) [inline service/src/PlatformTopo.cpp]: Use move semantics: mkstemp, make a temp file, write the lscpu contents to it, then move it to cache_file_name.

File: service/src/PlatformTopo.cpp [hunk 4]
```diff
@@ -423,31 +430,45 @@ namespace geopm
...
-            if (pclose(pid)) {
-                unlink(cache_file_name.c_str());
-                throw Exception("PlatformTopo::create_cache(): Could not pclose lscpu command: ",
-                         errno ? errno : GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
-            }
+        int err = stat(cache_file_name.c_str(), &cache_stat);
+        if (err) {
+            throw Exception("PlatformTopoImp::create_cache(): stat failure:",
+                     errno ? errno : GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+        }
```
Comment (bgeltz) [inline service/src/PlatformTopo.cpp]: This seems OK the way it was: if stat() had an RC of 0, don't regenerate the file.

**Comment by bgeltz** [comment]
Info about shlex parsing option "posix": https://docs.python.org/3.7/library/shlex.html#parsing-rules

**Review by cmcantalupo — DISMISSED** [review]

**Review by cmcantalupo — DISMISSED** [review]

File: scripts/geopmpy/launcher.py [hunk 5]
```diff
@@ -507,14 +507,13 @@ def run(self, stdout=sys.stdout, stderr=sys.stderr):
         if is_geopmctl:
             # Need to set OMP_NUM_THREADS to 1 in the env before the run
             self.config.set_omp_num_threads(1)
-            geopm_pid = subprocess.Popen(geopm_argv, env=self.environ(),
-                                         stdout=popen_stdout, stderr=popen_stderr,
-                                         shell=True)
+            geopm_pid = subprocess.Popen(shlex.split(geopm_argv), env=self.environ(),
```
Comment (dannosliwcd) [inline scripts/geopmpy/launcher.py]: Alternative to calling `shlex.split`, we can drop the `' '.join(...)` on line 482

File: scripts/geopmpy/plotter.py [hunk 6]
```diff
@@ -20,7 +20,7 @@
 
 try:
     with open(os.devnull, 'w') as FNULL:
-        subprocess.check_call("python -c 'import matplotlib.pyplot'", stdout=FNULL, stderr=FNULL, shell=True)
+        subprocess.check_call(['python', '-c',  "'import matplotlib.pyplot'"], stdout=FNULL, stderr=FNULL)
```
Comment (dannosliwcd) [inline scripts/geopmpy/plotter.py]: This probably always fails on systems that don't have an unversioned `python` symlink (vs `python3`).

**Review by dannosliwcd — DISMISSED** [review]
Two non-blocking suggestions. Current state also looks good.

**Review by dannosliwcd — APPROVED** [review]

**Pull request merged**